solr/src/main/java/dk/dbc/opensearch/solr/FilterQuery.java

```diff
@@ -0,0 +1,148 @@
+public class FilterQuery {
+
+    public static List<FlatQuery> from(FlatQuery query) {
+        return new FilterQuery().extractFrom(query).getFilterQueries();
+    }
+
+    private final Map<String, FlatQueryAndNot> filterQueriesByName;
+    private final List<FlatQuery> filterQueries;
+
+    private FilterQuery() {
+        this.filterQueriesByName = new HashMap<>();
+        this.filterQueries = new ArrayList<>();
+    }
+
+    private FlatQueryAndNot filterQueryFor(String filterQueryName) {
+        return filterQueriesByName.computeIfAbsent(filterQueryName, n ->
+                                           {
+                                               FlatQueryAndNot replace = new FlatQueryAndNot();
+                                               filterQueries.add(replace);
+                                               return replace;
+                                           });
+    }
+
+    private FilterQuery extractFrom(FlatQuery query) {
+        if (query instanceof FlatQueryAndNot) {
+            // extract only from top level and nodes
+            extract((FlatQueryAndNot) query);
+        } else if (query instanceof FlatQueryOr) {
+            extractOrOnTopNode((FlatQueryOr) query);
+        }
+        return this;
+    }
+
+    public List<FlatQuery> getFilterQueries() {
+        return filterQueries;
+    }
+
+    private void extract(FlatQueryAndNot flatQuery) {
+        for (Iterator<FlatQuery> iterator = flatQuery.ands().iterator() ; iterator.hasNext() ;) {
+            FlatQuery element = iterator.next();
+            String filterQueryName = findFilterQueryName(element);
+            if (filterQueryName != null) {
+                FlatQueryAndNot replace = filterQueryFor(filterQueryName);
+                iterator.remove();
+                replace.addAnd(element);
+            }
+        }
+    }
+
+    /**
+     * Clear top or node, and move all to new filterQuery or node
+     *
+     * @param flatQuery top node
+     */
+    private void extractOrOnTopNode(FlatQueryOr flatQuery) {
+        String filterQueryName = findFilterQueryName(flatQuery);
+        if (filterQueryName != null) {
+            FlatQueryOr filterQuery = new FlatQueryOr();
+            filterQueries.add(filterQuery);
+            filterQuery.ors().addAll(flatQuery.ors());
+            flatQuery.ors().clear();
+        }
+    }
+
+    private static String findFilterQueryName(FlatQuery query) {
+        if (query instanceof FlatQueryNested) {
+            return ( (FlatQueryNested) query ).getFilterQueryName();
+        } else if (query instanceof FlatQuerySearch) {
+            return ( (FlatQuerySearch) query ).getFilterQueryName();
+        } else if (query instanceof FlatQueryOr) {
+            Iterator<FlatQuery> i = ( (FlatQueryOr) query ).ors().iterator();
+            return allTheSameFilter(i);
+        } else if (query instanceof FlatQueryAndNot) {
+            Iterator<FlatQuery> i = ( (FlatQueryAndNot) query ).ands().iterator();
```

Comment:
Introducer lokal variabel så du ikke behøves at caste 3 forskellige steder
```suggestion
            FlatQueryAndNot andNotQuery = (FlatQueryAndNot) query
            Iterator<FlatQuery> i = andNotQuery.ands().iterator();
```